File: services/api/app/permissions.py
import enum


class Permission(str, enum.Enum):
    SYSTEM_ADMIN = "system_admin"
    MANAGE_HOUSEHOLDS = "manage_households"
    MANAGE_HOUSEHOLD_MEMBERS = "manage_household_members"
    MANAGE_PARENTAL_CONTROLS = "manage_parental_controls"
    MANAGE_ROLE_PROFILES = "manage_role_profiles"
    SET_PARENT_PIN = "set_parent_pin"
    SUBJECT_TO_PARENTAL_CONTROLS = "subject_to_parental_controls"


ROOT_ADMIN_PERMISSIONS = {perm.value for perm in Permission}

DEFAULT_PROFILE_DEFINITIONS: dict[str, dict[str, object]] = {
    "parent": {
        "name": "Parent",
        "permissions": {
            Permission.MANAGE_HOUSEHOLD_MEMBERS.value,
            Permission.MANAGE_PARENTAL_CONTROLS.value,
            Permission.SET_PARENT_PIN.value,
        },
    },
    "child": {
        "name": "Child",
        "permissions": {Permission.SUBJECT_TO_PARENTAL_CONTROLS.value},
    },
    "adult": {
        "name": "Adult",
        "permissions": set(),
    },
    "household_admin": {
        "name": "Household Administrator",
        "permissions": {
            Permission.MANAGE_HOUSEHOLD_MEMBERS.value,
            Permission.MANAGE_PARENTAL_CONTROLS.value,
            Permission.MANAGE_ROLE_PROFILES.value,
            Permission.SET_PARENT_PIN.value,
        },
    },
}

File: apps/web/src/lib/permissions.ts
import type { PermissionName, RoleProfileSummary, User } from "@/types";

export function hasPermission(user: User | null | undefined, permission: PermissionName): boolean {
  if (!user) return false;
  if (user.is_root_admin) return true;
  return user.permissions.includes(permission);
}

export function isChildProfile(user: User | null | undefined): boolean {
  return hasPermission(user, "subject_to_parental_controls");
}

export function canManageMembers(user: User | null | undefined): boolean {
  return hasPermission(user, "manage_household_members");
}

export function canManageParentalControls(user: User | null | undefined): boolean {
  return hasPermission(user, "manage_parental_controls");
}

export function canSetParentPin(user: User | null | undefined): boolean {
  return hasPermission(user, "set_parent_pin");
}

export function canManageRoleProfiles(user: User | null | undefined): boolean {
  return hasPermission(user, "manage_role_profiles");
}

export function formatRoleProfiles(profiles: RoleProfileSummary[]): string {
  return profiles.map((profile) => profile.name).join(", ");
}

export function findRoleProfileBySlug(profiles: RoleProfileSummary[], slug: string): RoleProfileSummary | undefined {
  return profiles.find((profile) => profile.slug === slug);
}
